onnxruntime/core/providers/xnnpack/detail/utils.cc

```diff
@@ -94,5 +244,207 @@ std::unique_ptr<IndexedSubGraph::MetaDef> FuseActivation(const Node& node, const
 
   return metadef;
 }
+
+bool IsQuantizedConv(QuantizedOpType quant_op_type) {
+  return (quant_op_type == QuantizedOpType::QLinearConv) ||
+         (quant_op_type == QuantizedOpType::QDQConv);
+}
+
+bool IsQuantizedMaxPool(QuantizedOpType quant_op_type) {
+  return (quant_op_type == QuantizedOpType::QLinearMaxPool) ||
+         (quant_op_type == QuantizedOpType::QDQMaxPool);
+}
+
+bool IsQuantizedAvgPool(QuantizedOpType quant_op_type) {
+  return (quant_op_type == QuantizedOpType::QlinearAvgPool) ||
+         (quant_op_type == QuantizedOpType::QDQAvgPool);
+}
+
+bool IsQuantizedSoftmax(QuantizedOpType quant_op_type) {
+  return (quant_op_type == QuantizedOpType::QDQSoftmax);
+}
+
+const onnx::TensorProto* GetQuantizationScale(const InitializedTensorSet& initializers,
+                                              const NodeUnitIODef& io_def) {
+  if (io_def.quant_param.has_value() == false) {
+    return nullptr;
+  }
+
+  onnx::TensorProto tensor_proto_ret;
+  const auto scale_name = io_def.quant_param->scale.Name();
+  auto it = initializers.find(scale_name);
+  if (it == initializers.cend()) {
+    return nullptr;
+  }
+  return it->second;
+}
+
+const onnx::TensorProto* GetQuantizationZeroPoint(const InitializedTensorSet& initializers,
+                                                  const NodeUnitIODef& io_def) {
+  if (!io_def.quant_param.has_value() || !io_def.quant_param->zero_point)
+    return nullptr;
+
+  const auto& zero_point_name = io_def.quant_param->zero_point->Name();
+  if (!Contains(initializers, zero_point_name)) {
+    return nullptr;
+  }
+
+  return initializers.at(zero_point_name);
+}
+
+// XNNPACK defined a few dtypes for quantized tensor, hence we can easily check if XNNPACK support it
+TensorQuantType GetTensorQuantType(const onnxruntime::NodeUnit& node_unit, int32_t io_index,
+                                   bool is_output, const onnxruntime::GraphViewer& graph_viewer) {
+  // we do not check the legality of io_index here
+  const NodeUnitIODef& iodef = is_output ? node_unit.Outputs()[io_index] : node_unit.Inputs()[io_index];
+  TensorQuantType datatype = TensorTypeInvalid;
+  int32_t input_type = 0;
+  if (!GetType(iodef.node_arg, input_type) || iodef.quant_param.has_value() == false) {
+    return datatype;
+  }
+
+  const InitializedTensorSet& initializers = graph_viewer.GetAllInitializedTensors();
```

Comment:
This needs to be limited to the constant initializers unless it will be dynamically provided at runtime.